timesketch/lib/analyzers/safebrowsing.py

```diff
@@ -0,0 +1,277 @@
+        """
+        # Exit ASAP if the API key is missing.
+        if not self._safebrowsing_api_key:
+            return 'Safe Browsing API requires an API key!'
+
+        query = (
+            '{"query": { "bool": { "should": [ '
+            '{ "exists" : { "field" : "url" }} ] } } }')
+
+        return_fields = ['url']
+
+        events = self.event_stream(
+            query_dsl=query,
+            return_fields=return_fields,
+        )
+
+        urls = {}
+
+        for event in events:
+            url = self._sanitize_url(event.source.get('url'))
+
+            if not url:
+                continue
+
+            urls.setdefault(url, []).append(event)
+
+        # Exit early if there are no URLs in the data set to analyze.
+        if not urls:
+            return 'No URLs to analyze.'
+
+        url_whitelisted = 0
+
+        url_whitelist = set(
+            interface.get_yaml_config(
+                self._URL_WHITELIST_CONFIG,
+            ),
+        )
+
+        if not url_whitelist:
+            domain_analyzer_whitelisted = current_app.config.get(
+                'DOMAIN_ANALYZER_WHITELISTED_DOMAINS',
+                [],
+            )
+            for domain in domain_analyzer_whitelisted:
+                url_whitelist.add('*.%s/*' % domain)
+
+        logging.info(
+            '{0:d} entries on the whitelist.'.format(len(url_whitelist)),
+        )
+
+        safebrowsing_platforms = current_app.config.get(
+            'SAFEBROWSING_PLATFORMS',
+            ['ANY_PLATFORM'],
+        )
+
+        safebrowsing_types = current_app.config.get(
+            'SAFEBROWSING_THREATTYPES',
+            ['MALWARE'],
+        )
+
+        lookup_urls = []
+
+        for url in urls:
+            if self._is_url_whitelisted(url, url_whitelist):
+                url_whitelisted += 1
+                continue
+
+            lookup_urls.append(url)
+
+        try:
+            safebrowsing_results = self._do_safebrowsing_lookup(
+                lookup_urls,
+                safebrowsing_platforms,
+                safebrowsing_types,
+            )
+        except requests.HTTPError:
+            return 'Couldn\'t reach the Safe Browsing API.'
+
+        for url, events in urls.items():
+            if url not in safebrowsing_results:
+                continue
+
+            safebrowsing_result = safebrowsing_results[url]
+            for event in events:
+                tags = ['google-safebrowsing-url']
+
+                threat_type = safebrowsing_result.get('threatType')
+
+                if threat_type:
+                    tags.append(
+                        'google-safebrowsing-%s' % threat_type.lower(),
+                    )
+
+                event.add_tags(tags)
+                event.add_attributes(
+                    {
+                        'google-safebrowsing-threat': safebrowsing_result,
+                    },
+                )
+                event.commit()
+
+        return (
```

Comment:
I also had some other comments , but I'll put them here.

No need to create any aggregations? Or views? Or an automatic story?

Reply:
Thanks, we might add something like these in another PR.